### PR TITLE
Preserving max output length for each sentence to avoid non-deterministic behavior using batching.

### PR DESCRIPTION
**Pull request by abdelrahmanbadawy**
(description of the change)

#### Pull Request Checklist ##
- [ ] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [ ] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [ ] System tests pass (`pytest test/system`)
- [ ] Passed code style checking (`./style-check.sh`)
- [ ] You have considered writing a test
- [ ] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [ ] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tdomhan** [comment]
there seems to be a python 3.4 incompatability:
```
E     File "/home/travis/build/awslabs/sockeye/sockeye/inference.py", line 1139
E       batch_translations = self._translate_nd(*self._get_inference_input(batch), max_output_lengths)
E                                                                                             ^
E   SyntaxError: only named arguments may follow *expression
```

File: sockeye/inference.py [hunk 1]
```diff
@@ -1132,7 +1133,10 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
             if rest > 0:
                 logger.debug("Extending the last batch to the full batch size (%d)", self.batch_size)
                 batch = batch + [batch[0]] * rest
-            batch_translations = self._translate_nd(*self._get_inference_input(batch))
+            max_output_lengths = []
+            for i in range(len(batch)):
+                max_output_lengths.append(self.models[0].get_max_output_length(len(batch[i])))
+            batch_translations = self._translate_nd(*self._get_inference_input(batch), max_output_lengths)
```
Comment (tdomhan) [inline sockeye/inference.py]: how about adding the creation of `max_output_lengths` directly to `_get_inference_input`? Also I'd directly create the NDArray and pass that instead of the list to _beam_search.

**Comment by tdomhan** [comment]
otherwise, looks good!

**Comment by tdomhan** [comment]
One more thing: please update the version number and add a small section to the CHANGELOG.

File: sockeye/image_captioning/inference.py [hunk 2]
```diff
@@ -141,7 +141,7 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
 
         return results
 
-    def _get_inference_input(self, image_paths: List[str]) -> Tuple[mx.nd.NDArray, int, List[Optional[constrained.RawConstraintList]]]:
+    def _get_inference_input(self, image_paths: List[str]) -> Tuple[mx.nd.NDArray, int, List[Optional[constrained.RawConstraintList]], mx.nd.NDArray]:
```
Comment (mjpost) [inline sockeye/image_captioning/inference.py]: Can you update the doc string to describe the additional return value?

File: sockeye/image_captioning/inference.py [hunk 3]
```diff
@@ -150,8 +150,13 @@ def _get_inference_input(self, image_paths: List[str]) -> Tuple[mx.nd.NDArray, i
         """
         ## TODO(bazzanil): support constraints
         raw_constraints = [None for x in range(self.batch_size)]  # type: List[Optional[constrained.RawConstraintList]]
+        max_output_lengths = []
```
Comment (mjpost) [inline sockeye/image_captioning/inference.py]: This should have a type declaration. Try:

    # type: List[int]

File: sockeye/inference.py [hunk 4]
```diff
@@ -1159,7 +1159,8 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
     def _get_inference_input(self,
                              trans_inputs: List[TranslatorInput]) -> Tuple[mx.nd.NDArray,
                                                                            int,
-                                                                           List[Optional[constrained.RawConstraintList]]]:
+                                                                           List[Optional[constrained.RawConstraintList]],
+                                                                           mx.nd.NDArray]:
```
Comment (mjpost) [inline sockeye/inference.py]: Please update doc string.

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,10 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.24]
+### Fixed
+- Preserving max output length for each sentence to avoid non-deterministic behavior using batching.
```
Comment (tdomhan) [inline CHANGELOG.md]: maybe phrase this as: ```...to allow for identical translation both with and without batching."```

File: sockeye/inference.py [hunk 6]
```diff
@@ -1193,7 +1196,10 @@ def _get_inference_input(self,
                 raw_constraints[j] = [data_io.tokens2ids(phrase, self.vocab_target) for phrase in
                                       trans_input.constraints]
 
-        return source, bucket_key, raw_constraints
+        # max output lengths extended to shape (batch_size * beam_size,)
+        max_out_lengths_extended = mx.nd.repeat(mx.nd.array(max_output_lengths, ctx=self.context, dtype='int32'), self.beam_size)
```
Comment (tdomhan) [inline sockeye/inference.py]: maybe just return the NDArray and do the actual `mx.nd.repeat` in `_beam_search` because it is only needed due to the fact that we have a beam. If we were to add something else, say greedy search, we could use the same method to create the input, while internally beam search would need to repeat the lengths.

**Comment by tdomhan** [comment]
thanks for iterating! :)

File: sockeye/image_captioning/inference.py [hunk 7]
```diff
@@ -141,17 +141,20 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
...
+        :return NDArray of images paths, bucket key, a list of raw constraint lists, an NDArray of maximum output lengths.
         """
         ## TODO(bazzanil): support constraints
         raw_constraints = [None for x in range(self.batch_size)]  # type: List[Optional[constrained.RawConstraintList]]
+        max_output_lengths = [] # type: List[int]
+        for i in range(len(image_paths)):
+            max_output_lengths.append(self.models[0].get_max_output_length(len(image_paths[i])))
         images = self.data_loader(image_paths, self.source_image_size)
-        return mx.nd.array(images), 0, raw_constraints
+        return mx.nd.array(images), 0, raw_constraints, mx.nd.array(max_output_lengths, ctx=self.context, dtype='int32')
```
Comment (tdomhan) [inline sockeye/image_captioning/inference.py]: actually here previously we passed the `0` as source length to `get_max_output_length`. Therefore you should do something like:
```
max_input_len = 0
max_output_lengths = mx.nd.array([self.models[0].get_max_output_length(max_input_len)] * len(image_paths[i]), ...)
return mx.nd.array(images), max_input_len, raw_constraints, max_output_lens
```

The actual value is not actually important as it is anyway overwritten through the `forced_max_output_len` flag when captioning, but just to stay consistent with the rest of of the code

File: sockeye/inference.py [hunk 8]
```diff
@@ -1535,6 +1544,7 @@ def _beam_search(self,
                 if mx.nd.sum(finished).asscalar() > 0:
                     break
             else:
+                finished = mx.nd.sign(finished + (mx.nd.cast(lengths.reshape(-1), 'int32') > max_output_lengths))
```
Comment (tdomhan) [inline sockeye/inference.py]: I think this check should actually be `lengths >= max_output_lengths`. For a single translation at the end of the for loop lengths will be exactly equal to max_output_length (because it starts at 1).

File: sockeye/image_captioning/inference.py [hunk 9]
```diff
@@ -128,8 +128,8 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
                 if self.source_root is not None:
                     path = os.path.join(self.source_root, path)
                 paths.append(path)
-            batch_translations = self._translate_nd(*self._get_inference_input(paths))
             # truncate to remove filler translations
+            batch_translations = self._translate_nd(*self._get_inference_input(paths))
```
Comment (fhieber) [inline sockeye/image_captioning/inference.py]: seems like an unnecessary change

Comment (tdomhan) [reply sockeye/image_captioning/inference.py]: the reordering with the comment? I agree.

File: sockeye/inference.py [hunk 10]
```diff
@@ -1363,6 +1369,9 @@ def _beam_search(self,
         lengths = mx.nd.ones((self.batch_size * self.beam_size, 1), ctx=self.context)
```
Comment (artemsok) [inline sockeye/inference.py]: can lengths be created with dtype='int32' not to be casted in L1373?

File: sockeye/inference.py [hunk 11]
```diff
@@ -1535,6 +1544,7 @@ def _beam_search(self,
                 if mx.nd.sum(finished).asscalar() > 0:
                     break
             else:
+                finished = mx.nd.sign(finished + (mx.nd.cast(lengths.reshape(-1), 'int32') >= max_output_lengths))
```
Comment (artemsok) [inline sockeye/inference.py]: matter of taste probably, but ```mx.nd.logical_or(finished, lengths.reshape(-1) >= max_output_lengths))``` might be a bit more readable

File: sockeye/inference.py [hunk 12]
```diff
@@ -1159,24 +1159,28 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
...
         """
 
         bucket_key = data_io.get_bucket(max(len(inp.tokens) for inp in trans_inputs), self.buckets_source)
         source = mx.nd.zeros((len(trans_inputs), bucket_key, self.num_source_factors), ctx=self.context)
         raw_constraints = [None for x in range(self.batch_size)]  # type: List[Optional[constrained.RawConstraintList]]
 
+        max_output_lengths = [] # type: List[int]
         for j, trans_input in enumerate(trans_inputs):
             num_tokens = len(trans_input)
+            max_output_lengths.append(self.models[0].get_max_output_length(num_tokens))
```
Comment (artemsok) [inline sockeye/inference.py]: just wondernig: could the self.models[i] return different max output lengths? if yes, should a max be taken over them?

File: sockeye/inference.py [hunk 13]
```diff
@@ -1363,6 +1369,9 @@ def _beam_search(self,
         lengths = mx.nd.ones((self.batch_size * self.beam_size, 1), ctx=self.context)
         finished = mx.nd.zeros((self.batch_size * self.beam_size,), ctx=self.context, dtype='int32')
 
+        # Extending max_output_lengths to shape (batch_size * beam_size,)
+        max_output_lengths = mx.nd.repeat(max_output_lengths, self.beam_size)
```
Comment (artemsok) [inline sockeye/inference.py]: what's the relation to ```max_output_length``` (w/o trailing s) now? 
would it be correct to set ```max_output_length = nd.max(max_output_lengths)``` before the repeat?

File: sockeye/inference.py (hunk 12 above)
Comment (tdomhan) [reply sockeye/inference.py]: load_models makes sure they don't. That said it is indeed not clear from the code and we should potentially refactor this.

File: sockeye/inference.py (hunk 11 above)
Comment (tdomhan) [reply sockeye/inference.py]: makes sense. I wasn't aware of the logical_or operator.

File: sockeye/inference.py (hunk 13 above)
Comment (tdomhan) [reply sockeye/inference.py]: yes, that's the case. Maybe we should indeed make this explicit the way you wrote it.

File: sockeye/inference.py (hunk 11 above)
Comment (fhieber) [reply sockeye/inference.py]: Please check if logical ops are part of mxnet1.2 already

File: sockeye/inference.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (abdelrahmanbadawy) [reply sockeye/inference.py]: That would cause the following error:
mxnet.base.MXNetError: Error in operator normalizefinishedhypotheses13__div0: [13:00:12] src/operator/contrib/../elemwise_op_common.h:123: Check failed: assign(&dattr, (*vec)[i]) Incompatible attr in node normalizefinishedhypotheses13__div0 at 1-th input: expected float32, got int32

I believe L1725: scores_accumulated / self.length_penalty(lengths),

File: sockeye/inference.py (hunk 11 above)
Comment (abdelrahmanbadawy) [reply sockeye/inference.py]: logical_or is not part of mxnet1.2 api installed through pip

Comment (fhieber) [reply sockeye/inference.py]: thanks for checking. we should switch to the logical ops only with mxnet 1.3 then. We have a few other places where these would be handy.

File: sockeye/inference.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/inference.py]: either way there has to be at least one casting because lengths is used as int type and as float type when computing the length penalty.

File: sockeye/inference.py [hunk 14]
```diff
@@ -1756,4 +1765,4 @@ def hybrid_forward(self, F, scores, finished, inactive, scores_accumulated, inf_
         # pad_dist. Shape: (batch*beam, vocab_size)
         pad_dist = F.concat(pad_id_scores, pad_dist)
         scores = F.where(finished + inactive, pad_dist, scores)
-        return scores
```
Comment (fhieber) [inline sockeye/inference.py]: missing newline at EOF

File: sockeye/inference.py [hunk 15]
```diff
@@ -1731,7 +1741,6 @@ def hybrid_forward(self, F, finished, inactive, lengths, best_word_indices):
         finished = ((best_word_indices == self.pad_id) + (best_word_indices == self.eos_id))
         return finished, finished_or_inactive, lengths
 
-
```
Comment (fhieber) [inline sockeye/inference.py]: pep8: 2 newlines between classes

File: sockeye/inference.py [hunk 16]
```diff
@@ -1164,24 +1164,28 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
...
-                bucket key, list of raw constraint lists.
+                bucket key, list of raw constraint lists,
+                NDArray of maximum output lengths.
         """
 
         bucket_key = data_io.get_bucket(max(len(inp.tokens) for inp in trans_inputs), self.buckets_source)
         source = mx.nd.zeros((len(trans_inputs), bucket_key, self.num_source_factors), ctx=self.context)
         raw_constraints = [None for x in range(self.batch_size)]  # type: List[Optional[constrained.RawConstraintList]]
 
+        max_output_lengths = [] # type: List[int]
```
Comment (fhieber) [inline sockeye/inference.py]: pep8: two spaces

File: sockeye/__init__.py [hunk 17]
```diff
@@ -11,4 +11,4 @@
 # express or implied. See the License for the specific language governing
 # permissions and limitations under the License.
 
-__version__ = '1.18.28'
+__version__ = '1.18.29'
```
Comment (fhieber) [inline sockeye/__init__.py]: missing newline

**Comment by tdomhan** [comment]
Resolved the version conflict. Waiting for Travis, then merging.
